Add text-to-speech functionality

ryanflach opened the pull request:
This PR completes base functionality of the CLI accepting input from the user and converting to speech. Additionally, documentation has been updated and completed.

Notes by file:
- `lib/text_talk.ex`
  - `start/2` is run automatically with `mix run`
  - Triggers output of a welcome message and runs the `get_input/0` function in the Speech module.
- `lib/text_talk/speech.ex`
  - Main functionality lives here
  - documentation is discarded for private methods, so it is only supplied for `get_input/0`
  - `process_input/1` splits the input string at '|' and sends to exit if the head of the created List is 'exit' - matching case insensitively. Otherwise the split input is sent to one of two functions.
  - `send_to_speech/1` has two variants - one with a clause for when there is only a head in the List, and one that processes a list with two items. Both send the appropriate args to `speak/2`
  - `speak/2` accepts two args and has a default for the second (voice). This function triggers the actual speech. It returns a tuple of `{:ok, line, voice}`, though this would only been seen in iex.
  - `exit_program/0` outputs a message and exits the program.

In `exit_program/0` I am using `System.halt(0)` to exit discretely (i.e., without warning output). I was originally attempting to use `exit/1` but was unable to find a means of doing so discretely. Because of the simplicity of this program (no other running processes) `System.halt(0)` seems fine, but I would love to know of a better way, if there is something more appropriate in this context.

ryanflach committed:
Update documentation

ryanflach committed:
Add TextTalk module to application for auto-run with mix run

ryanflach committed:
Remove old test structure

ryanflach committed:
Add text-to-speech functionality

matthewrpacker commented:
@ryanflach Nice work on the documentation, method names, and variable names.  Your code is clean and easy to follow.  I have zero experience with Elixir, but I found that because of your naming and clarity, I was able to understand the functionality.  I don't have any technical input.  Nice work!

saylerb commented on lib/text_talk/speech.ex:
```diff
@@ -0,0 +1,45 @@
+  end
+
+  defp speak(line, voice \\ "Junior") do
+    System.cmd "say", ["-v#{voice}", line]
```
Is the `System.cmd "say"` part of macOS or *nix?  If it's not cross-platform, maybe think about indicating that in the readme. I haven't wrote any elixir before but it's very easy to follow what you've written. Good job!

ryanflach replied:
@saylerb Good thought. I don't believe it works on Windows. I'll update the readme to reflect that.

Jbern16 commented:
those are some sexy lamdas mate

Jbern16 commented on lib/text_talk/speech.ex:
```diff
@@ -0,0 +1,45 @@
+
+  defp process_input(input) do
+    line = String.split(input, "|")
+           |> Enum.map(fn(x) -> String.trim(x) end)
```
mmmmm

ryanflach committed:
Update documentation to reflect inability of use on Windows

mattr- commented on lib/text_talk/speech.ex:
```diff
@@ -0,0 +1,45 @@
+  end
+
+  defp process_input(input) do
+    line = String.split(input, "|")
```
You could refactor this to

``` elixir
input
|> String.split("|")
|> Enum.map(&String.trim/1)
```

ryanflach replied:
@mattr- I really like this refactor. Would `trim(&1)` be the same as `trim/1`? They both seem to achieve the same functionality. Could you elaborate some on why you'd use `trim/1` over `trim(&1)`?

mattr- replied:
Sort of. The use of `trim/1` is so that you can make the reference to the `trim/1` method in the `String` module. Elixir (and erlang as well, I believe) refer to functions by their arity and so this is explicitly stating what function should be called for the `map` operation.

ryanflach replied:
That makes sense. Thank you.

mattr- commented:
So, there are some things that you can do with this to improve the code and make it a bit more idiomatic Elixir and take advantage of the language's pattern matching and tail recursion facilities.

You can pipe the return value of `IO.gets` into `process_input`, like so:

`````` elixir
IO.gets "<Your prompt>"
|> process_input```

which lets you take advantage of tail recursion by calling `get_input` at the end of itself. So, putting the whole thing together:

```elixir
def get_input do
  IO.puts "Note: To specify a voice, pass it as a second argument after '|'"
  IO.puts "For example, 'hello | Boing' would say hello in the Boing voice.\n\n"
  IO.gets "What would you like me to repeat ('exit' to quit)?\n"
  |> process_input(line)
  get_input
end
``````

This lets you remove the calls to `get_input` that you've scattered all throughout the rest of the other methods.

You can also take advantage of Elixir's pattern matching capabilities to do some crazy stuff with your function definitions. Instead of what you have for `send_to_speech` now, you could do something like:

``` elixir
defp send_to_speech(["exit"]), do: exit_program
defp send_to_speech([text]), do: speak(text)
defp send_to_speech([text, voice | _]), do: speak(text, voice)
```

I'll let you dissect this one and figure it out :smiley:

ryanflach commented:
@mattr- Thanks for the great comments. Because `IO.gets` has an arity of 2, I had to put it into parentheses in order to utilize that refactor (it was thinking my pipe operator was the second arg), but I really like the refactor. Removing the extra calls to get_input makes it a bit cleaner.

I'm still very new to Elixir, and pattern matching has taken the longest to sink in, so I hadn't thought of using it here. I'm interpreting the suggestions surrounding `send_to_speech` as follows:

```
defp send_to_speech(["exit"]), do: exit_program
```

This checks the head of the list, and if it is "exit" calls the `exit_program` function. This would allow me to remove the conditional from `process_input/1`. Because the conditional was handling case-insensitivity, I've modified the chain for assignment to `line`, ultimately being able to remove that variable assignment altogether:

```
defp process_input(input) do
    input
      |> String.downcase
      |> String.split("|")
      |> Enum.map(&String.trim(&1))
      |> send_to_speech
  end
```

To maintain functionality of exit working even if they've supplied a voice, I'm thinking of modifying the `send_to_speech` for exit to:

```
  defp send_to_speech(["exit" | _]), do: exit_program
```

```
defp send_to_speech([text]), do: speak(text)
```

I interpret this similarly to `send_to_speech(["exit"])`, only it is not looking for a specific value, just ensuring that the list only has one item.

```
defp send_to_speech([text, voice | _]), do: speak(text, voice)
```

I interpret this as saying that you want to assign variables to the values in the list as `text` and `voice`, split the list after the second argument, ultimately ignoring the tail by means of `_`, then send the values of `text` and `voice` as arguments to `speak/2`. 

Like I said, I'm new to Elixir, so I really appreciate your comments!

ryanflach committed:
Refactor functions to utilize pattern matching

mattr- commented:
> I had to put it into parentheses in order to utilize that refactor

Idiomatic Elixir indicates that you should always use parenthesis unless you have no arguments (https://github.com/levionessa/elixir_style_guide#syntax from the Elixir Style Guide for reference)

> To maintain functionality of exit working even if they've supplied a voice, I'm thinking of modifying the send_to_speech for exit to:
> 
> `defp send_to_speech(["exit" | _]), do: exit_program`

Yup. That's right!

> I interpret this similarly to send_to_speech(["exit"]), only it is not looking for a specific value, just ensuring that the list only has one item.

Correct

>  interpret this as saying that you want to assign variables to the values in the list as text and voice, split the list after the second argument, ultimately ignoring the tail by means of _, then send the values of text and voice as arguments to speak/2.

Also correct.

ryanflach commented:
@mattr- You were a huge help. Thank you.

zackforbing commented:
being completely unfamiliar with Elixir, I think this code looks super clean and readable. just going on context clues, I'm guessing `defp` is a private method? I was also interested in the test until I got to the assertion. your documentation on such a simple app is, as always, incredibly thorough. keep up the good work.
